shared/utils/relabel-ids.py

```diff
@@ -73,236 +256,115 @@ def create_xccdf_id_to_cce_id_mapping(xccdftree):
-                        # This is the required XCCDF 'type' for <xccdf:Value> derived
-                        # from OVAL variable 'datatype' and mapping above
-                        reqxccdftype = oval_to_xccdf_datatype_constraints[ovalvartype]
-                        # Compare the actual value of 'type' of <xccdf:Value> with the requirement
-                        if xccdfvartype != reqxccdftype:
-                            # If discrepancy is found, issue a warning
-                            sys.stderr.write(
-                                "Warning: XCCDF 'type' of \"%s\" value does "
-                                "not meet the XCCDF value 'type' to OVAL "
-                                "variable 'datatype' export matching "
-                                "constraint! Got: \"%s\", Expected: \"%s\". "
-                                "Resetting it! Set 'type' of \"%s\" "
-                                "<xccdf:value> to '%s' directly in the XCCDF "
-                                "content to dismiss this warning!" %
-                                (ovalvarid, xccdfvartype, reqxccdftype,
-                                 ovalvarid, reqxccdftype)
-                            )
-                            # And reset the 'type' attribute of such a <xccdf:Value> to the required type
-                            xccdfvar.attrib['type'] = reqxccdftype
+    # Verify if <xccdf:Value> 'type' to corresponding OVAL variable 'datatype' export matching constraint:
+    #
+    # http://csrc.nist.gov/publications/nistpubs/800-126-rev2/SP800-126r2.pdf#page=30&zoom=auto,69,313
+    #
+    # is met. Also correct the 'type' attribute of those <xccdf:Value> elements where necessary in
+    # order the produced content to meet this constraint.
+    #
+    # To correct the constraint we use simpler approach - prefer to fix 'type' attribute of <xccdf:Value>
+    # rather than 'datatype' attribute of the corresponding OVAL variable since there might be additional
+    # OVAL variables, derived from the affected OVAL variable, and in that case we would need to fix the
+    # 'datatype' attribute in each of them.
+
+
+    # Define the <xccdf:Value> 'type' to OVAL variable 'datatype' export matching constraints mapping
+    # as specified in Table 16 of XCCDF v1.2 standard:
+    # http://csrc.nist.gov/publications/nistpubs/800-126-rev2/SP800-126r2.pdf#page=30&zoom=auto,69,313
+    #
+    indexed_xccdf_values = ssgcommon.map_elements_to_their_ids(xccdftree, ".//{%s}Value" % (xccdf_ns))
+
+    # Loop through all <external_variables> in the OVAL document
+    ovalextvars = ovaltree.findall(".//{%s}external_variable" % oval_ns)
+    if ovalextvars is None:
+        return
+
+    for ovalextvar in ovalextvars:
+        # Verify the found external variable has both 'id' and 'datatype' set
+        if 'id' not in ovalextvar.attrib or 'datatype' not in ovalextvar.attrib:
+            msg = "Invalid OVAL <external_variable> found."
```

Comment:
Some error messages can't be very specific, because not enough info is available. However, higher level of the code adds info about the OVAL file, where the error originates.